### PR TITLE
Add DestructArray and DestructArrayAndReassign operations

### PR DESCRIPTION
**Pull request by amarekano**
This implementation currently does not support the following JS constructs:
* ~Nested arrays (i.e. [[a,b],c] = arr)~ To be handled during compilation
* ~holes (i.e [a,,b] = arr)~
* ~rest element (i.e. [a, ...b] = arr)~

### CODE REVIEW

File: Sources/Fuzzilli/Core/CodeGenerators.swift [hunk 1]
```diff
@@ -371,6 +371,31 @@ public let CodeGenerators: [CodeGenerator] = [
         b.reassign(target, to: val)
     },
 
+    CodeGenerator("DestructArrayGenerator", input: .iterable) { b, arr in
+        // TODO: We need a way to expose the length of arr so that element selection is accurate
```
Comment (saelo) [inline Sources/Fuzzilli/Core/CodeGenerators.swift:375]: This'd be quite complex to do, and I don't think we really need it: 
```javascript
let arr = [1,2,3];
let [x, y] = arr;
// x = 1, y = 2
let [v1, v2, v3, v4] = arr;
// v4 = undefined
```
seems to work fine.

Maybe just update this comment to say something like "we could improve this logic by tracking the lengths of arrays and using that, but it's not strictly necessary since array destructuring allows the number of variables to be different from the length of the array"

File: Sources/Fuzzilli/FuzzIL/Operations.swift [hunk 2]
```diff
@@ -622,6 +622,31 @@ class Reassign: Operation {
     }
 }
 
+/// Destructs an array into n output variables
+class DestructArray: Operation {
+    let indices: [Int]
+    let hasRestElement: Bool
+    
+    init(indices: [Int], hasRestElement: Bool) {
+        self.hasRestElement = hasRestElement
```
Comment (saelo) [inline Sources/Fuzzilli/FuzzIL/Operations.swift:631]: `indices` must be sorted here, correct? Maybe add an `assert` to verify that? I guess it must also not have duplicates, although that might be a bit harder to assert (maybe convert to a `Set`, then compare against that?), so if it's more than one short-ish line, it's probably fine to just add a comment for that property

File: Sources/Fuzzilli/Lifting/JavaScriptLifter.swift [hunk 3]
```diff
@@ -115,6 +115,44 @@ public class JavaScriptLifter: Lifter {
...
+                        indices = op.indices
+                        hasRestElement = op.hasRestElement
+                    case let op as DestructArrayAndReassign:
+                        outputs = instr.inputs.dropFirst().map({ $0.identifier })
+                        indices = op.indices
+                        hasRestElement = op.hasRestElement
+                    default:
+                        return arrayPattern
+                }
+                assert(indices.count == outputs.count)
```
Comment (saelo) [inline Sources/Fuzzilli/Lifting/JavaScriptLifter.swift:137]: Can we copy this assertion into the operation constructor as well?

File: Sources/Fuzzilli/Lifting/JavaScriptLifter.swift [hunk 4]
```diff
@@ -115,6 +115,44 @@ public class JavaScriptLifter: Lifter {
...
+                                hasRestElement ? arrayPattern.append(" ...\(outputs[outputIndex])") : arrayPattern.append(" \(outputs[outputIndex])")
+                            } else {
+                                arrayPattern.append("\(outputs[Int(outputIndex)]),")
+                            }
+                            outputIndex += 1
+                        } else {
+                            arrayPattern.append(" ,")
+                        }
+                    }
+                }
```
Comment (saelo) [inline Sources/Fuzzilli/Lifting/JavaScriptLifter.swift:151]: Can we simplify this a bit and make it O(N) (not O(N^2), not that it really matters though). Something like
```swift
var lastIndex = 0
for (index, output) in zip(indices, outputs) {
    let skipped = index - lastIndex
    lastIndex = index
    let dots = index == indices.last! && hasRestElement ? "..." : ""
    arrayPattern += String(repeating: ",", count: skipped) + dots + output + ","
}
```
maybe?

File: Sources/Fuzzilli/Protobuf/program.pb.swift [hunk 5]
```diff
@@ -132,777 +132,799 @@ public struct Fuzzilli_Protobuf_Instruction {
   /// The operation is either encoded as an index, referring to the nth operation
   /// (so that shared operations are also only present once in the protobuf), or
   /// as one of the many concrete Operation messages.
-  public var inouts: [UInt32] = []
+  public var inouts: [UInt32] {
```
Comment (saelo) [inline Sources/Fuzzilli/Protobuf/program.pb.swift:135]: Are all of these changes due to a newer version of the Swift protobuf compiler? I'd slightly prefer to do that in a separate PR if possible, but it's fine if that's too much hassle (e.g. getting an older version to work)

File: Sources/Fuzzilli/FuzzIL/Operations.swift [hunk 6]
```diff
@@ -622,6 +622,31 @@ class Reassign: Operation {
     }
 }
 
+/// Destructs an array into n output variables
+class DestructArray: Operation {
+    let indices: [Int]
```
Comment (saelo) [inline Sources/Fuzzilli/FuzzIL/Operations.swift:627]: Do we want the OperationMutator to be able to modify these (i.e. turn one index into another)? I think that could make sense, although it'll need to ensure the indices are still sorted and contain no duplicates

File: Sources/Fuzzilli/Lifting/FuzzILLifter.swift [hunk 7]
```diff
@@ -220,6 +220,50 @@ public class FuzzILLifter: Lifter {
...
+                    } else {
+                        arrayPattern.append(",")
+                    }
+                }
+            }
+            w.emit("[\(arrayPattern)] <- DestructArray \(input(0))")
+
+        case let op as DestructArrayAndReassign:
+            var arrayPattern: String = ""
+            if op.indices.count > 0 {
```
Comment (saelo) [inline Sources/Fuzzilli/Lifting/FuzzILLifter.swift:247]: Maybe similar code changes as in the JSLifter?

File: Sources/Fuzzilli/Core/CodeGenerators.swift [hunk 8]
```diff
@@ -371,6 +371,31 @@ public let CodeGenerators: [CodeGenerator] = [
...
+        // Pick random indices
+        var indices: [Int] = []
+        for idx in 0..<Int.random(in: 0..<5) {
```
Comment (saelo) [inline Sources/Fuzzilli/Core/CodeGenerators.swift:378]: Maybe add a short comment about how these random values are determined (5 and 0.7). I think 5 makes sense since it's the default max length of array literals (is it?). Then the probability should probably be selected so that on average, we define 2-3 new variables?

**Review by saelo — COMMENTED** [review]
Nice! The main thing I'm still wondering about is whether we really need the `AndReassign` variant of this. On the one hand, it most likely gets converted early on in the JS pipeline into a destruct and regular reassignments, so probably doesn't really cover any new code. On the other hand, compiling a destruct + reassign to a normal destruct and a bunch of reassignments is kind of inefficient, and so if that happens frequently enough, it's probably worth keeping the additional instruction for that. WDYT?

File: Sources/Fuzzilli/Core/CodeGenerators.swift (hunk 1 above)
Comment (amarekano) [reply Sources/Fuzzilli/Core/CodeGenerators.swift:375]: Actually, I can just drop the comment since it's strictly not necessary that we cap the array length.

File: Sources/Fuzzilli/Lifting/JavaScriptLifter.swift (hunk 4 above)
Comment (amarekano) [reply Sources/Fuzzilli/Lifting/JavaScriptLifter.swift:151]: Nice! I learn a new swift construct with each PR I send :D
I'll update this logic.

File: Sources/Fuzzilli/Protobuf/program.pb.swift (hunk 5 above)
Comment (amarekano) [reply Sources/Fuzzilli/Protobuf/program.pb.swift:135]: Yeah, I've updated my protobuf compiler so there's a whole load of this changes.

**Comment by amarekano** [comment]
> The main thing I'm still wondering about is whether we really need the AndReassign variant of this.
Let me compare the bytecode generate from `destruct + reassign` and `destruct with multiple assignments`. Perhaps we can take a decision using that data.

File: Sources/Fuzzilli/Lifting/JavaScriptLifter.swift (hunk 3 above)
Comment (amarekano) [reply Sources/Fuzzilli/Lifting/JavaScriptLifter.swift:137]: I think you mean adding this in ProgramBuilder right? In the Operation constructor we won't have access to the output variables list.

File: Sources/Fuzzilli/Protobuf/program.pb.swift (hunk 5 above)
Comment (saelo) [reply Sources/Fuzzilli/Protobuf/program.pb.swift:135]: Hmm are you sure that's the latest release? For me I have
```
> protoc-gen-swift --version
protoc-gen-swift 1.18.0
```
which seems to be the latest version according to https://github.com/apple/swift-protobuf

File: Sources/Fuzzilli/Lifting/JavaScriptLifter.swift (hunk 3 above)
Comment (saelo) [reply Sources/Fuzzilli/Lifting/JavaScriptLifter.swift:137]: Ah yeah right, nevermind. The Operation specifies `numOutput` and that is asserted to be equal to the number of outputs in the Instruction, so we don't need a separate check.

File: Sources/Fuzzilli/Protobuf/program.pb.swift (hunk 5 above)
Comment (amarekano) [reply Sources/Fuzzilli/Protobuf/program.pb.swift:135]: AH! Turns out I picked the last listed tag to compile from which was (1.9.0) and that appears after 1.18.0. I'll recompile and rebase my PRs :(

File: Sources/Fuzzilli/Core/CodeGenerators.swift (hunk 8 above)
Comment (amarekano) [reply Sources/Fuzzilli/Core/CodeGenerators.swift:378]: The max length for array generation that we currently have is 10 elements, Should we continue with 5 as the upper limit for array destruct?

Comment (saelo) [reply Sources/Fuzzilli/Core/CodeGenerators.swift:378]: Right ok, yeah maybe still leave it at 5 then, since that's the expected number of elements in an array if the max is 10. WDYT?

Comment (amarekano) [reply Sources/Fuzzilli/Core/CodeGenerators.swift:378]: Yeah that works for me too

**Comment by amarekano** [comment]
I think we should support the `AndReassign` variant for a couple of reasons: the first being that this op generates more interesting bytecode compared to just using simple reassignments. Here's two programs that I've used to test this:

```js
const v4 = [15,30,"Hello"];
let v5 = 1000;
[v5] = v4;

//VS

const v4 = [15,30,"Hello"];
let v5 = 1000;
v5 = v4[0];

```
The bc from the two samples are linked below:
[bc_with_destruct_array.log](https://github.com/googleprojectzero/fuzzilli/files/7573017/bc_with_destruct_array.log)
[bc_without_destruct_array.log](https://github.com/googleprojectzero/fuzzilli/files/7573024/bc_without_destruct_array.log)

The second reason is that handling `restElement` assignments would become tricky. Thinking off the top of my head, I imagine you'd need to introduce a loop (or some slice operation) that would simulate the same reassignment, which would change the semantics of the program.

**Comment by amarekano** [comment]
Wrt to nested arrays in destruct operations, I think we can handle this during compilation using temporary variables. For example, if we have the following program:
```js
let [x,[a,b]] = [1,[2,3]]
```
we can re-write this during compilation to:

```js
let [x, _temp0] = [1,[2,3]]
let [a,b] = _temp0
```
Diffing the bytecode dumps from the [orginal](https://github.com/googleprojectzero/fuzzilli/files/7576412/orginal_bc.log) and [modified sample](https://github.com/googleprojectzero/fuzzilli/files/7576413/rewrite_bc.log) above doesn't reveal a significant difference so I think we should be fine on that front as well.

All other review comments have been addressed so we should be good to go here. Let me know if you'd like me to incorporate any other changes or improvements.

File: Sources/Fuzzilli/Mutators/OperationMutator.swift [hunk 9]
```diff
@@ -117,6 +117,22 @@ public class OperationMutator: BaseInstructionMutator {
             newOp = BinaryOperation(chooseUniform(from: allBinaryOperators))
         case is ReassignWithBinop:
             newOp = ReassignWithBinop(chooseUniform(from: allBinaryOperators))
+        case let op as DestructArray:
+            var newIndices: [Int] = []
```
Comment (saelo) [inline Sources/Fuzzilli/Mutators/OperationMutator.swift:121]: I think I'd prefer if these did smaller changes. Also I'm not sure that this currently works? The number of outputs might now be different than the number of indices. How about something like:
```swift
var newIndices = Set(op.indices)
if newIndices.count > 0 {
    newIndices.remove(newIndices.randomElement()!)
    while newIndices.count != op.indices.count {
        newIndices.insert(Int.random(in: 0..<10))
    }
}
newOp = ... (newIndices.sorted(), ...)
...
```

Comment (saelo) [reply Sources/Fuzzilli/Mutators/OperationMutator.swift:121]: I think it's ok to use a larger maximum number (10) here than in the generator

File: Sources/Fuzzilli/Core/CodeGenerators.swift [hunk 10]
```diff
@@ -393,6 +393,32 @@ public let CodeGenerators: [CodeGenerator] = [
         b.reassign(target, to: val)
     },
 
+    CodeGenerator("DestructArrayGenerator", input: .iterable) { b, arr in
+        // Fuzzilli generated arrays can have a length ranging from 0 to 10 elements,
+        // We want to ensure that 1) when destructing arrays we are within this length range
```
Comment (saelo) [inline Sources/Fuzzilli/Core/CodeGenerators.swift:398]: Nit: maybe "when destructing arrays we are usually within", because we might still generate arrays with only 1 or 2 elements

**Review by saelo — COMMENTED** [review]
Cool thanks, that makes sense, so let's keep both variants!

File: Sources/Fuzzilli/Core/ProgramBuilder.swift [hunk 11]
```diff
@@ -1309,6 +1309,16 @@ public class ProgramBuilder {
         perform(Reassign(), withInputs: [output, input])
     }
 
+    @discardableResult
+    public func destruct(_ input: Variable, selecting indices: [Int], hasRestElement: Bool = false) -> [Variable] {
+        let outputs = perform(DestructArray(indices: indices, hasRestElement: hasRestElement), withInputs: [input]).outputs
+        return Array(outputs)
+    }
+
+    public func destructAndReassign(_ candidates: [Variable], atIndices indices: [Int], from input: Variable, hasRestElement: Bool = false) {
```
[the remark this answers is not in the document]
Comment (amarekano) [reply Sources/Fuzzilli/Core/ProgramBuilder.swift:1318]: thoughts on renaming this function to just `destruct` and updating the signature as you've suggested?

File: Sources/Fuzzilli/Protobuf/operations.pb.swift [hunk 12]
```diff
@@ -950,6 +950,34 @@ public struct Fuzzilli_Protobuf_Reassign {
...
 }
 
+public struct Fuzzilli_Protobuf_DestructArray {
+  // SwiftProtobuf.Message conformance is added in an extension below. See the
+  // `Message` and `Message+*Additions` files in the SwiftProtobuf library for
+  // methods supported on all messages.
+
+  public var indices: [Int32] = []
+
+  public var hasRestElement_p: Bool = false
```
[the remark this answers is not in the document]
Comment (amarekano) [reply Sources/Fuzzilli/Protobuf/operations.pb.swift:960]: I think this generated by the swift proto complier (which I've now correctly upgraded to 1.18.0, thanks for pointing that out :)) and it appears to be a standard representation for `bool` fields. We've also got this notation in BeginClassDef:
```swift
case .beginClassDefinition(let p):
            op = BeginClassDefinition(hasSuperclass: p.hasSuperclass_p,
```

File: Sources/FuzzilliCli/CodeGeneratorWeights.swift [hunk 13]
```diff
@@ -70,6 +70,8 @@ let codeGeneratorWeights = [
     "ReassignWithBinopGenerator":               25,
     "DupGenerator":                             2,
     "ReassignmentGenerator":                    25,
+    "DestructArrayGenerator":                   2,
+    "DestructArrayAndReassignGenerator":        7,
```
[the remark this answers is not in the document]
Comment (amarekano) [reply Sources/FuzzilliCli/CodeGeneratorWeights.swift:74]: Not really, these are just arbitrary values at the moment. Happy to change them to something else if you've got suggestions.

[the remark this answers is not in the document]
Comment (amarekano) [reply Sources/FuzzilliCli/CodeGeneratorWeights.swift:74]: I've updated these to now have the same value of 7 which is the average of weights from `ArrayGenerator` and `ArrayWithSpreadGenerator`. Again, this is purely empirical :)

[the remark this answers is not in the document]
Comment (saelo) [reply Sources/FuzzilliCli/CodeGeneratorWeights.swift:74]: Sounds good, thanks!

File: Sources/Fuzzilli/Core/ProgramBuilder.swift (hunk 11 above)
[the remark this answers is not in the document]
Comment (saelo) [reply Sources/Fuzzilli/Core/ProgramBuilder.swift]: sure, that works as well!
